### PR TITLE
Copying settings to secure screens doesn't warn when all add-ons are disabled

### LINKED ISSUE #8274: Copying settings to secure screens gives a warning about addons even when they are all disabled.
### Steps to reproduce:

1. Ensure that you have one or more addons installed, but all disabled.
1. Open the General settings dialog (`NVDA+control+g`).
1. Ensure that "Use NVDA on the Windows logon screen (requires administrator privileges)" is checked
1. Press the "Use currently saved settings on the logon and other secure screens (requires administrator privileges)" button.

### Expected behavior:
No warnings

### Actual behavior:
The following dialog is shown: "Warning  dialog  Add-ons were detected in your user settings directory. Copying these to the system profile could be a security risk. Do you still wish to copy your settings?"

### Discussion:
There is no way to enable addons from the log on screen, so copying these addons is not useful. This should be updated so only enabled addons are copied, and the warning message lists which addons will be copied to make it easier for the user to make an informed decision. When there are no enabled addons, none should be copied, and there should be no warning message.

### System configuration:
NVDA version:  next-15060,4709bfa7

NVDA Installed or portable:
Installed

Windows version:
Windows 10 Version 1709

### PR DESCRIPTION
**Pull request by nvdaes**
### Link to issue number:
Fixes #8274

### Summary of the issue:
NVDA shouldn't give a warning when users choose to copy settings for secure screens, if all add-ons are disabled.

### Description of how this pull request fixes the issue:
We check if globalVars.appArgs.disableAddons is False (not globalVars.appArgs.disableAddons) to see if all add-ons are disabled. In this case, the for loop to check if directories contained in the user configuration folder aren't empty, in order to ask confirmation before copying settings, is not performed.

### Testing performed:
Unit tests and run NVDA master with this pull request from source.
Due to an error with some libraries on the local computer, this snapshot hasn't been installed for a real test.

### Known issues with pull request:
None.
Anyway, this could be improved:
1. Checking if there are or not running add-ons, to copy just them and show their name to the user.
2. Remove old plugins and drivers.
3. Don't warn if they aren't running add-ons though globalVars.appArgs.disableAddons is True (for instance, if they have been disabled individually).

### Change log entry:

* Changes
NVDA doesn't warn before copying settings for secure screens when all add-ons have been disabled.

### CODE REVIEW

**Comment by leonardder** [comment]
Thanks for your pr.

What worries me here is that this creates an inconsistency. When add-ons are force disabled, this code does not copy them at all, while as when some are disabled and others are enabled, all addons are still copied regardless of their state.

**Comment by nvdaes** [comment]
Yes, I thought about this too, and my idea was that someone could use the feature of disabling add-ons precissely as a temporary way of indicating that they shouldn't be copied, without disabling it the next time NVDA is restarted.
Anyway, for me old plugins should be removed and then a for loop between running add-ons should let to select individual add-ons to copy.

**Comment by Brian1Gaff** [comment]
This may seem a silly question, but if using settings on secure screens, are 
there any circumstances where anything but a synth add on is actually of any 
consequence to the working of nvda on that screen?
 If so why not disable all add ons which are not synths anyway and make the 
copy smaller and presumably less of a security issue in any case.
 Brian

**Comment by nvdaes** [comment]
@Brian1Gaff , not just add-ons containing syntesizer drivers could take effect in secure screens and may be copied.

**Comment by nvdaes** [comment]
Now addonHandler.getRunningAddons() method is used instead.
For each running add-on, its add-on summary is appended to a list named runningAddons. The warning is shown just when there are running add-ons, without looking to app modules, global plugins and drivers, since probably they will be removed and the whole config directory with its subdirs are copied now.
This could be improved, as mentioned in this related issue, where the possibility of selecting what add-ons will be copied to system configuration is requested:
https://github.com/nvaccess/nvda/issues/6305

I have installed this snapshot of NVDA creating a launcher, and have made tests disabling various add-ons and all.
For me results are acceptable. The summary of the running add-ons is shown when the warning is presented.
I think it's better to show summaries and not names, since they are intended to be the name to display and are shown in the list of the add-on manager.
@leonardder and others, if you are interested, please review this or provide more feed-back.
Thanks

**Comment by nvdaes** [comment]
Last commit not tested yet.

File: .github/issue_template.md [hunk 1]
```diff
@@ -1,12 +1,34 @@
+> Below you will find NVDA's Github issue template. You might consider swapping to the preview tab in order
```
Comment (leonardder) [inline .github/issue_template.md]: I don't understand why this is in the diff, could you please merge most recent master?

File: source/config/__init__.py [hunk 2]
```diff
@@ -246,6 +246,11 @@ def _setSystemConfig(fromPath):
 		if curSourceDir==fromPath:
 			curDestDir=toPath
 		else:
+			if os.path.relpath(curSourceDir,fromPath) == "addons":
+				import addonHandler
+				subDirs[:] = [addon.name for addon in addonHandler.getRunningAddons()]
+			else:
+				subDirs[:] = subDirs
```
Comment (leonardder) [inline source/config/__init__.py]: I don't think this is necessary.

File: source/gui/settingsDialogs.py [hunk 3]
```diff
@@ -703,16 +704,17 @@ def makeSettings(self, settingsSizer):
...
-					_("Add-ons were detected in your user settings directory. Copying these to the system profile could be a security risk. Do you still wish to copy your settings?"),
-					# Translators: The title of the warning dialog displayed when trying to copy settings for use in secure screens.
-					_("Warning"),wx.YES|wx.NO|wx.ICON_WARNING,self
-				)==wx.NO:
-					return
-				break
+		runningAddons = []
+		for addon in addonHandler.getRunningAddons():
+			runningAddons.append(addon.manifest['summary'])
+		if len(runningAddons)>0:
```
Comment (leonardder) [inline source/gui/settingsDialogs.py]: You can easily change this into something like:
`if next(addonHandler.getRunningAddons(), None):`

**Review by LeonarddeR — CHANGES_REQUESTED** [review]

**Comment by nvdaes** [comment]
Thanks, @leonardder.
I've merged master and simplified the code of gui following your excellent suggestion.
About the config file, are you sure? According to my tests in Python console, I think that the assignment subDirs[:]=subDirs is necessary. Otherwise, when curSourceDir is "addons", the assignment can be changed causing issues, that is, I think the subDir[:] needs to be restored when curSourceDir is not "addons".
But, if you are sure, I will change it.
About addonHandler, it's imported in other function of config/__init__.py, but not at module level. I don't know if this could be simplified importing it just once.
Regards.

**Comment by nvdaes** [comment]
Now results are as follows:
* When trying to copy settings, list of summaries for running add-ons is shown in the message box requesting confirmation.
* The user config folder is copied, as previously, with all containing subfolders, i.e., non running add-ons are copied, but they can't be enabled on secure screens. This has the dfollowing drawback: it's not useful, the same as other temporary files contained in the user config folder, or possibly dictionaries for synthesizers, which can remain in NVDA even when corresponding voices have been uninstalled, for instance. The advantage could be if NVDA could include a feature for reverting configuration by copying options from system config.
* Drivers intended to be used for testing, like appModules, globalPlugins, etc., are copied too. Since NVDA doesn't have a proper way to provide a name, description or other tasks of management for this, maybe better not to copy them.
* I think this shouldn't be merged until the issue about selecting individual add-ons to copy is addressed. Though disabling individual add-ons before copying settings could be a way of selecting them.

Thanks

**Comment by josephsl** [comment]
Hi,

Can you rebase this branch to master if possible? Or, we may need to start over from scratch due to #6275 and so many changes since then.

Thanks.

**Comment by nvdaes** [comment]
@josephsl, for now I see conflicts to fix in various files. I will get a look at this.

**Comment by nvdaes** [comment]
@josephsl, you can test changes. Just a few lines need to be changed in gui/settingsDialogs after merging master into this branch.
scons tests have passed and NVDA runs well.
Thanks

**Comment by josephsl** [comment]
Hi,

I think telling users that these add-ons are currently enabled might help, since it now checks if any add-ons are running i.e. enabled. Something to the effect of:

The following add-ons are currently enabled.

Eventually I think a dialog to choose individual add-ons to copy would build on top of this change.

Also, when copying add-ons, copy only add-ons that are running if users said "yes" to the warning, but then again, I think a dialog described above may make things a bit easier.

Thanks.

**Comment by nvdaes** [comment]
@Josephsl commented:

> Eventually I think a dialog to choose individual add-ons to copy would build on top of this change.

This was suggested by @leonardder in issue #6305.

 > Also, when copying add-ons, copy only add-ons that are running if users said "yes" to the warning.

Maybe discussed, taking account as a drawback that pickle file will contain info for add-ons not copied, and that if the user configuration could be restored to the system configuration, the add-ons not copied couldn't be enabled later.

**Comment by josephsl** [comment]
Hi,

Are we ready to get this PR going again?

Thanks.

**Comment by feerrenrut** [comment]
We have decided to close this PR. We would like to see much deeper user control for how add-ons are copied to the system profile. A GUI should be introduced, allowing (but strongly discouraging) users to select which add-ons to use on the system profile. No add-ons should be selected by default.

**Comment by nvdaes** [comment]
OK, thanks.

2020-07-03 10:05 GMT+02:00, Reef Turner <notifications@github.com>:

**Comment by cary-rowen** [comment]
@nvdaes 
Do you have any new progress on this PR? Regarding #6305, of course it is not just this closed PR. I hope to see new progress.
Thanks.

**Comment by nvdaes** [comment]
@cary-rowen, no, there is not any progress for now. Please see issue #11006

**Pull request closed without merging**